src/algebra/char_p/mixed_char_zero.lean

```diff
@@ -0,0 +1,452 @@
+such that the quotient `R ⧸ I` has caracteristic `p`.
+
+**Remark:** For `p = 0`, `mixed_char R 0` is a meaningless definition as `R ⧸ ⊥ ≅ R` has by
+definition always characteristic zero.
+One could require `(I ≠ ⊥)` in the definition, but then `mixed_char R 0` would mean something
+like `ℤ`-algebra of extension degree `≥ 1` and would be completely independent from
+whether something is a `ℚ`-algebra or not (e.g. `ℚ[X]` would satisfy it but `ℚ` wouldn't).
+-/
+class mixed_char_zero (p : ℕ) : Prop :=
+[to_char_zero : char_zero R]
+(char_p_quotient : ∃ (I : ideal R), (I ≠ ⊤) ∧ char_p (R ⧸ I) p)
+
+namespace mixed_char_zero
+
+/--
+Reduction to `p` prime: When proving any statement `P` about mixed characteristic rings we
+can always assume that `p` is prime.
+-/
+lemma reduce_to_p_prime {P : Prop} :
+  (∀ p > 0, (mixed_char_zero R p → P)) ↔
+  (∀ (p : ℕ), (nat.prime p → mixed_char_zero R p → P)) :=
+begin
+  split,
+  { intros h q q_prime q_mixed_char,
+    exact h q (nat.prime.pos q_prime) q_mixed_char },
+  { intros h q q_pos q_mixed_char,
+    rcases q_mixed_char.char_p_quotient with ⟨I, ⟨hI_ne_top, hI_char⟩⟩,
+
+    -- Krull's Thm: There exists a prime ideal `P` such that `I ≤ P`
+    rcases ideal.exists_le_maximal I hI_ne_top with ⟨M, ⟨hM_max, h_IM⟩⟩,
+    resetI, -- make `hI_char : char_p (R ⧸ I) q` an instance.
+
+    let r := ring_char (R ⧸ M),
+    have r_pos : r ≠ 0 :=
+    begin
+      have q_zero := congr_arg (ideal.quotient.factor I M h_IM) (char_p.cast_eq_zero (R ⧸ I) q),
+      simp only [map_nat_cast, map_zero] at q_zero,
+      apply ne_zero_of_dvd_ne_zero (ne_of_gt q_pos),
+      exact (char_p.cast_eq_zero_iff (R ⧸ M) r q).mp q_zero
+    end,
+    have r_prime : nat.prime r :=
+      or_iff_not_imp_right.1 (char_p.char_is_prime_or_zero (R ⧸ M) r) r_pos,
+    apply h r r_prime,
+    haveI : char_zero R := q_mixed_char.to_char_zero,
+    exact
+    ⟨begin
+      use M,
+      split,
+      exact hM_max.ne_top,
+      refine ring_char.of_eq rfl
+    end⟩ }
+end
+
+/--
+Reduction to `I` prime ideal: When proving statements about mixed characteristic rings,
+after we reduced to `p` prime, we can assume that the ideal `I` in the definition is maximal.
+-/
+lemma reduce_to_maximal_ideal {p : ℕ} (hp : nat.prime p) :
+  (∃ (I : ideal R), (I ≠ ⊤) ∧ char_p (R ⧸ I) p) ↔
+  (∃ (I : ideal R), (I.is_maximal) ∧ char_p (R ⧸ I) p) :=
+begin
+  split,
+  { intro g,
+    rcases g with ⟨I, ⟨hI_not_top, hI⟩⟩,
+
+    -- Krull's Thm: There exists a prime ideal `M` such that `I ≤ M`.
+    rcases ideal.exists_le_maximal I hI_not_top with ⟨M, ⟨hM_max, hM⟩⟩,
+
+    use M,
+    split,
+    exact hM_max,
+    { cases char_p.exists (R ⧸ M) with r hr,
+      convert hr,
+      resetI, -- make `hr : char_p (R ⧸ M) r` an instance.
+
+      have r_dvd_p : r ∣ p :=
+      begin
+        rw ←char_p.cast_eq_zero_iff (R ⧸ M) r p,
+        convert congr_arg (ideal.quotient.factor I M hM) (char_p.cast_eq_zero (R ⧸ I) p)
+      end,
+      rw eq_comm,
```

Comment:
Very much a question of taste but `symmetry` is a dedicated tactic for doing exactly this:
```suggestion
      symmetry,
```